### ISSUE TITLE
Normalization for tabulardata produces nan for constant columns

### ISSUE THREAD

**Issue by jhmenke** [fastai/fastai#889]
**Describe the bug**
While creating a TabularDataset from a pandas DataFrame, the normalization will fail if some columns have constant values. This is an edge case where some variable might be constant for most samples (e.g., all samples in the test set), but change in some occasions.

https://github.com/fastai/fastai/blob/c93bbf7fa4e92ef12540698824071397d130cc4c/fastai/tabular/data.py#L35

This line will turn some columns of `self.conts` to nan if `stds` is 0. It can be avoided by replacing the line with this code:

```
self.conts = (self.conts - means[None])
inds = stds != 0.
self.conts[:, inds] /= stds[inds]
```

It will only perform the division for nonzero standard deviations. The other `self.conts` entries will be 0. anyway due `self.conts - means[None]` being equal for constant columns.

**Comment by sgugger** [comment]
Shouldn't such a variable be treated as categorical then?

**Comment by jhmenke** [comment]
it could be a continuous variable.
an example for a technical process would a a temperature measurement, which could be 20.0 in a controlled environment but vary slightly in some samples (e.g. 19.9 or 20.1). If a small data set is used it could appear to be a constant but vary in other data sets.

**Comment by bhoomit** [comment]
This can also happen in a case where one is trying something out on a subset of data, which is not representative of the entire dataset.

**Comment by jhmenke** [comment]
i tried this as well, but discarded this solution: keep in mind that with a small epsilon all values in the column will become huge (e.g. 1e12 or whatver), which is a big contrast to the other values. for example, my validation loss would be something like 2384589345 for a few training iterations because of the small epsilon.

**Comment by jph00** [comment]
We may change this again before we get to the tabular lesson in a couple of weeks. We'll do some more testing.
